### PR TITLE
Add rust API documentation to the matrix module

### PR DESCRIPTION
**Pull request by agunde406**
Signed-off-by: Andrea Gunderson <agunde@bitwise.io>

### CODE REVIEW

File: libsplinter/src/matrix.rs [hunk 1]
```diff
@@ -46,20 +64,67 @@ impl Envelope {
...
+    /// * `message` - the bytes of the messages
+    ///
+    /// If the send failed, a MatrixSendError will be returned.
     fn send(&self, id: String, message: Vec<u8>) -> Result<(), MatrixSendError>;
 }
 
+/// MatrixReceiver trait abstracts out receiving messages from a connection handler without
+/// requiring knowledge adding and removing connections, and sending messages
 pub trait MatrixReceiver: Clone + Send {
+    /// Attempt to receive a message from the connection handler. This function will block until
```
Comment (peterschwarz) [inline libsplinter/src/matrix.rs:104]: Also would be sure to note here that the message will come from the first ready connection detected.

File: libsplinter/src/matrix.rs [hunk 2]
```diff
@@ -12,32 +12,50 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+//! Traits that abstract specific functionality of connection handler.
+//!
+//! A component that acts as a connection handler peforms the following operations
```
[the remark this answers is not in the document]
Comment (agunde406) [reply libsplinter/src/matrix.rs:17]: changed

File: libsplinter/src/matrix.rs [hunk 3]
```diff
@@ -12,32 +12,50 @@
...
+//!   - add: Adds a connection to the underlying connection handler
+//!   - remove: Removes a connection from the connection handler
+//!   - send: Send a message over a specific connection
+//!   - recv: Receive a message from any connection in the connection handler
```
[the remark this answers is not in the document]
Comment (agunde406) [reply libsplinter/src/matrix.rs:21]: added

File: libsplinter/src/matrix.rs (hunk 1 above)
Comment (agunde406) [reply libsplinter/src/matrix.rs:104]: added

File: libsplinter/src/matrix.rs [hunk 4]
```diff
@@ -12,54 +12,120 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+//! Traits that abstract specific functionality of connection group.
```
Comment (ltseeley) [inline libsplinter/src/matrix.rs:15]: Should this be "a connection group."?

File: libsplinter/src/matrix.rs [hunk 5]
```diff
@@ -12,54 +12,120 @@
...
+//!
+//! A component that acts as a connection group peforms the following operations
+//!   - add: Adds a connection to the underlying connection group
+//!   - remove: Removes a connection from the connection group
+//!   - send: Send a message over a specific connection
+//!   - recv: Receive a message from any ready connection in the connection group
+//!
+//! The following Matrix traits abstract out these different operations. For example an
+//! and networking component may only be interested in send and receiving messages. That component
```
Comment (ltseeley) [inline libsplinter/src/matrix.rs:24]: I assume this should be "For example, a networking component..."? Also s/send/sending/

File: libsplinter/src/matrix.rs [hunk 6]
```diff
@@ -12,54 +12,120 @@
...
-/// Wrapper around payload to include connection id
+/// Wrapper around a payload to include connection ID
 #[derive(Debug, Default, PartialEq)]
 pub struct Envelope {
     id: String,
     payload: Vec<u8>,
 }
 
 impl Envelope {
+    /// Create a new Envelope that will be sent over a connection
```
Comment (ltseeley) [inline libsplinter/src/matrix.rs:43]: I think this should be changed to "Creates ..." to match the style guidelines (same with "Returns ..." for the payload method)

File: libsplinter/src/matrix.rs [hunk 7]
```diff
@@ -12,54 +12,120 @@
...
 /// MatrixLifeCycle trait abstracts out adding and removing connections to a
-/// connection handler without requiring knowledge about sending or receiving messges.
+/// connection group without requiring knowledge about sending or receiving messges.
 pub trait MatrixLifeCycle: Clone + Send {
+    /// Add a connection to the connection group with a connection ID
+    ///
+    /// # Arguments
+    ///
+    /// * `connection` - A boxed connection that will be passed to the connection group.
+    /// * `id` - a unique connection ID that will be used to sent messages over a connection.
```
Comment (ltseeley) [inline libsplinter/src/matrix.rs:72]: s/sent/send/

File: libsplinter/src/matrix.rs [hunk 8]
```diff
@@ -12,54 +12,120 @@
...
     pub fn take_payload(self) -> Vec<u8> {
         self.payload
     }
 }
 
 /// MatrixLifeCycle trait abstracts out adding and removing connections to a
-/// connection handler without requiring knowledge about sending or receiving messges.
+/// connection group without requiring knowledge about sending or receiving messges.
 pub trait MatrixLifeCycle: Clone + Send {
+    /// Add a connection to the connection group with a connection ID
```
Comment (ltseeley) [inline libsplinter/src/matrix.rs:67]: s/Add/Adds/

File: libsplinter/src/matrix.rs [hunk 9]
```diff
@@ -12,54 +12,120 @@
...
+    /// # Arguments
+    ///
+    /// * `id` - the unique connection ID for the connection that should be removed
+    ///
+    /// If the remove failed, a MatrixRemoveError will be returned.
     fn remove(&self, id: &str) -> Result<Box<dyn Connection>, MatrixRemoveError>;
 }
 
+/// MatrixSender trait abstracts out sending messages through a connection group without
+/// requiring knowledge adding and removing connections, and receiving messages
```
Comment (ltseeley) [inline libsplinter/src/matrix.rs:88]: I think this can use some rewording (same for receiver trait)

File: libsplinter/src/matrix.rs [hunk 10]
```diff
@@ -12,54 +12,120 @@
...
 pub trait MatrixSender: Clone + Send {
+    /// Send a message over the specified connection.
+    ///
+    /// # Arguments
+    ///
+    /// * `id` - the unique connection ID the message should be sent over
+    /// * `message` - the bytes of the messages
```
Comment (ltseeley) [inline libsplinter/src/matrix.rs:95]: s/messages/message/

File: libsplinter/src/matrix.rs [hunk 11]
```diff
@@ -12,54 +12,120 @@
...
+    ///
+    /// If the send failed, a MatrixSendError will be returned.
     fn send(&self, id: String, message: Vec<u8>) -> Result<(), MatrixSendError>;
 }
 
+/// MatrixReceiver trait abstracts out receiving messages from a connection group without
+/// requiring knowledge adding and removing connections, and sending messages
 pub trait MatrixReceiver: Clone + Send {
+    /// Attempt to receive a message from the connection group. This function will block until
```
Comment (ltseeley) [inline libsplinter/src/matrix.rs:104]: s/Attempt/Attempts/ (same for recv_timeout)

File: libsplinter/src/matrix.rs [hunk 12]
```diff
@@ -12,54 +12,120 @@
...
+    /// there is a Envelope to receive. The message will come from the first ready connection
```
Comment (ltseeley) [inline libsplinter/src/matrix.rs:105]: We should probably put things like "Envolope" between ticks so it shows up as a reference to code in the docs.

File: libsplinter/src/matrix.rs [hunk 13]
```diff
@@ -12,54 +12,120 @@
...
+    ///
+    /// If successful, returns an Envelope containing the payload and the connection ID of the
+    /// the sender. Otherwise, returns a MatrixRecvTimeoutError.
     fn recv_timeout(&self, timeout: Duration) -> Result<Envelope, MatrixRecvTimeoutError>;
 }
 
+/// MatrixShutdown trait abstracts out shutting down a connection group without requiring
+/// knowledge of the other connection group operations.
 pub trait MatrixShutdown: Clone + Send {
+    /// Function to notify the underlying connection group to shutdown
```
Comment (ltseeley) [inline libsplinter/src/matrix.rs:128]: s/Function to notify/Notifies/

**Comment by vaporos** [comment]
The 'connection group' terminology feels awkward in places (though if we moved forward on renaming matrix it would make sense).

**Pull request merged**